### PR TITLE
Use a single socket for InterfaceChoice.Default

### LINKED ISSUE #330: Should not bind to all interfaces by default
Since the changes in #270/#188, the default behavior is to listen & broadcast on all available interfaces. Since we work with multicast, every other interface can see the broadcast. We end up reading our own sent packets which results in broadcast storms that overwhelm the host/network.

This is _especially_ problematic on busy Docker hosts. In these setups each container will add another interface/bridge to the box which amplifies the issue.

This behavior can be overridden by explicitly choosing an interface when instantiating but this new default seems dangerous.

### PR DESCRIPTION
**Pull request by bdraco**
Fixes #330 

When using multiple sockets with multi-cast, the outgoing
socket's responses could be read back on the incoming socket,
which leads to duplicate processing and could fill up the
incoming buffer before it could be processed.

This behavior manifested with error similar to
`OSError: [Errno 105] No buffer space available`

By using a single socket with InterfaceChoice.Default
we avoid this case.

### CODE REVIEW

**Comment by bdraco** [comment]
@jjlawren Can you give this a shot with `https://www.home-assistant.io/integrations/zeroconf/#default_interface` set to `True`?

**Comment by bdraco** [comment]
Verified .All still works
```
[pid 19654] recvfrom(8, "\0\0\204\0\0\0\0\2\0\0\0\33\4_hap\4_tcp\5local\0\0\f\0"..., 8966, 0, {sa_family=AF_INET, sin_port=htons(5353), sin_addr=inet_addr("169.254.236.67")}, [16]) = 1460
[pid 19654] select(18, [8 9 10 12 13 14 16 17], [], [], {5, 0}) = 1 (in [8], left {4, 999997})
[pid 19654] recvfrom(8, "\0\0\204\0\0\0\0\2\0\0\0\33\4_hap\4_tcp\5local\0\0\f\0"..., 8966, 0, {sa_family=AF_INET, sin_port=htons(5353), sin_addr=inet_addr("127.0.0.1")}, [16]) = 1460
[pid 19654] select(18, [8 9 10 12 13 14 16 17], [], [], {5, 0}) = 1 (in [8], left {4, 999998})
[pid 19654] recvfrom(8, "\0\0\204\0\0\0\0\2\0\0\0\33\4_hap\4_tcp\5local\0\0\f\0"..., 8966, 0, {sa_family=AF_INET, sin_port=htons(5353), sin_addr=inet_addr("192.168.107.5")}, [16]) = 1460
[pid 19654] select(18, [8 9 10 12 13 14 16 17], [], [], {5, 0}) = 1 (in [8], left {4, 999998})
[pid 19654] recvfrom(8, "\0\0\204\0\0\0\0\2\0\0\0\33\4_hap\4_tcp\5local\0\0\f\0"..., 8966, 0, {sa_family=AF_INET, sin_port=htons(5353), sin_addr=inet_addr("172.17.0.1")}, [16]) = 1460
[pid 19654] select(18, [8 9 10 12 13 14 16 17], [], [], {5, 0}) = 1 (in [8], left {4, 999998})
```

**Comment by bdraco** [comment]
```
[pid 20649] select(14, [12 13], [], [], {5, 0}) = 1 (in [12], left {4, 999997})
[pid 20649] recvfrom(12, "\0\0\0\0\0\0\0\3\0\0\0\1\4_hap\4_tcp\5local\0\0\f\0"..., 8966, 0, {sa_family=AF_INET, sin_port=htons(5353), sin_addr=inet_addr("192.168.107.116")}, [16]) = 158
```
and .Default

**Comment by jjlawren** [comment]
Ran some basic tests against this branch. Everything seems to be working as I'd expect it to with far less traffic.

**Comment by bdraco** [comment]
I've got it running on 4 systems and all the buffer errors no longer happen :)

File: zeroconf/__init__.py [hunk 1]
```diff
@@ -2978,8 +2994,9 @@ def close(self) -> None:
             if not self.unicast:
                 self.engine.del_reader(cast(socket.socket, self._listen_socket))
                 cast(socket.socket, self._listen_socket).close()
-            for s in self._respond_sockets:
-                self.engine.del_reader(s)
+            if self.unicast or self.interfaces is not InterfaceChoice.Default:
```
Comment (jstasiak) [inline zeroconf/__init__.py:2997]: Can we extract the `self.unicast or self.interfaces is not InterfaceChoice.Default` into a function or something? I think `_read_from_respond_sockets` could be a good name for it.

**Review by jstasiak — APPROVED** [review]
I swear one of these days I need to make the logic that choses, creates and configures sockets into a testable unit, sometimes I can't wrap my head around this logic anymore.

Can you update the commit message with some details and motivation behind this? I'd like those to be readily available through the commit history rather than needed to be found on GitHub.

Looks good otherwise. :)

File: zeroconf/__init__.py (hunk 1 above)
Comment (bdraco) [reply zeroconf/__init__.py:2997]: I can make it a property that gets setup at `__init__`

Comment (jstasiak) [reply zeroconf/__init__.py:2997]: Yup, works for me.

**Comment by bdraco** [comment]
Retesting..

**Comment by bdraco** [comment]
Sadly thats not explicit enough since it needs to pickup the unicast cast.

File: zeroconf/__init__.py (hunk 1 above)
Comment (bdraco) [reply zeroconf/__init__.py]: Went with `self.multi_socket` since all the conditionals would be `not self.single_socket`

**Comment by bdraco** [comment]
Retested `python3 3881 root    6u     IPv4             330088      0t0       UDP *:mdns ` 👍 

Working as expected

**Comment by jstasiak** [comment]
Nice!

**Comment by bdraco** [comment]
Thanks, would you mind pushing a release so we can merge https://github.com/home-assistant/core/pull/48302 ?

**Comment by jstasiak** [comment]
Oh yeah, sure. There's an issue with the typing that I haven't spotted before merging, would you mind having a look?

```
zeroconf/__init__.py:2335: error: List item 0 has incompatible type "Optional[socket]"; expected "socket"
```

(from https://travis-ci.org/github/jstasiak/python-zeroconf/jobs/764291256)

**Comment by bdraco** [comment]
Looking..

**Comment by bdraco** [comment]
https://github.com/jstasiak/python-zeroconf/pull/333

Also I can open a PR to switch to GH actions if you like

**Comment by jstasiak** [comment]
Oh, I'll take care of that in a moment, I have a template for that already, but thank you for the offer.

**Comment by jstasiak** [comment]
@bdraco Released in 0.29.0!

**Comment by bdraco** [comment]
> @bdraco Released in 0.29.0!

Thank you 🙏🏻

**Pull request merged**